### PR TITLE
Cosine annealing warm restarts (T_mult) cross-dataset (code change required)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Standard CosineAnnealingLR uses a fixed period T_max. CosineAnnealingWarmRestarts (SGDR) restarts the cosine cycle at increasing intervals (T_0, T_0*T_mult, T_0*T_mult^2, ...), allowing the model to escape sharp minima repeatedly while the restart intervals grow to permit finer convergence. This technique has shown strong results in vision and NLP tasks. We hypothesize it will improve over fixed-period cosine on DrivAerML and TandemFoil, especially with longer-period restarts (T_mult=2–3) that give progressively longer exploration windows.

## Code Change Required

**You must add `--cosine-t-mult` support to `target/icml2026/train.py` before running.**

Add the argument parser entry:
```python
parser.add_argument('--cosine-t-mult', type=int, default=1,
                    help='T_mult multiplier for CosineAnnealingWarmRestarts (1 = standard CosineAnnealingLR)')
```

Replace the scheduler construction block (find where `CosineAnnealingLR` is instantiated) with:
```python
from torch.optim.lr_scheduler import CosineAnnealingLR, CosineAnnealingWarmRestarts

if hasattr(args, 'cosine_t_mult') and args.cosine_t_mult > 1:
    scheduler = CosineAnnealingWarmRestarts(optimizer, T_0=args.cosine_t_max, T_mult=args.cosine_t_mult)
else:
    scheduler = CosineAnnealingLR(optimizer, T_max=args.cosine_t_max)
```

## Instructions

After implementing the code change above, run the following 8 experiments. **Do not hardcode SENPAI_TIMEOUT_MINUTES or SENPAI_MAX_EPOCHS.** Use `--epochs 999` and let the environment control the budget.

**Golden configs (baselines for each dataset):**

DAM base:
```
python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
```

AF base:
```
python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999
```

TF base:
```
python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999
```

**Run assignments (all use `--wandb_group warm-restarts`):**

- GPU 0: TF base + `--cosine-t-max 10 --cosine-t-mult 2`
- GPU 1: AF base + `--cosine-t-max 5 --cosine-t-mult 2`
- GPU 2: DAM base + `--cosine-t-max 5 --cosine-t-mult 2`
- GPU 3: DAM base + `--cosine-t-max 10 --cosine-t-mult 2`
- GPU 4: DAM base + `--cosine-t-max 15 --cosine-t-mult 2`
- GPU 5: DAM base + `--cosine-t-max 30 --cosine-t-mult 2`
- GPU 6: DAM base + `--cosine-t-max 5 --cosine-t-mult 3`
- GPU 7: DAM base + `--cosine-t-max 10 --cosine-t-mult 3`

Run all 8 in parallel, one per GPU. Report `val_primary/surface_rel_l2_pct` for DrivAerML, `val_primary/surface_mse` for AirfRANS, and `val_primary/surface_pressure_mae` for TandemFoil.

## Baseline

**DrivAerML:** `val_primary/surface_rel_l2_pct = 4.619%` (PR #2691, epoch 256, 4L/512d/8H, AdamW lr=5e-4, T_max=30)
Reproduce: `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

**AirfRANS:** `val_primary/surface_mse = 0.001236` (PR #2828, epoch 349, 2L/256d/4H, AdamW lr=7e-4, gc=1.0, WD=1e-2, T_max=5)
Reproduce: `cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999`

**TandemFoil:** `val_primary/surface_pressure_mae = 30.10` (PR #2810, epoch 157, 3L/192d/3H, Lion lr=1.25e-4, gc=1.0, WD=1e-2, T_max=10)
Reproduce: `cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT mugen:

## Results

### Summary

Warm restarts (CosineAnnealingWarmRestarts with T_mult) show **strong wins on AirfRANS and TandemFoil** but **degrade DrivAerML** across all tested configurations.

### AirfRANS — T_0=5, T_mult=2 — NEW BEST ✓

| Metric | Baseline | Warm Restarts | Change |
|--------|----------|---------------|--------|
| `val_primary/surface_mse` | 0.001236 (ep349) | **0.000371** (ep221) | **-70.0%** |
| `test_primary/surface_mse` | — | 0.001188 (ep351) | — |
| External target | 0.0043 | **0.000371** | 91.4% below |

- **Per-channel val MSE at best (ep221):** Ux=1.5e-05, Uy=1.1e-06, nut=1.3e-06, p=0.001466
- **W&B run:** 8x579r3n (351 epochs, 360-min budget)
- **Command:** `CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --cosine-t-mult 2 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent mugen --wandb-name "mugen/af-warm-restarts-t5-m2" --wandb-group "warm-restarts"`
- **Analysis:** Growing restart periods (5→10→20→40→... batches) allow the model to repeatedly escape shallow minima early on, then converge more precisely as periods lengthen. The 70% val improvement is dramatic. Test metric at final epoch (0.001188) is worse than best val (0.000371) because test is evaluated at ep351 — model likely regressed from the ep221 trough or was at an LR peak. **Checkpointing at best val is critical to capture the true performance.**

### TandemFoil — T_0=10, T_mult=2 — NEW BEST ✓

| Metric | Baseline | Warm Restarts | Change |
|--------|----------|---------------|--------|
| `val_primary/surface_pressure_mae` | 30.10 (ep157) | **25.46** (ep109) | **-15.4%** |
| `test_primary/surface_pressure_mae` | — | 36.29 (ep135) | — |

- **Per-split val MAE at best (ep109):** geom_camber_cruise=13.91, single_in_dist=24.17, re_rand=26.93, geom_camber_rc=36.82
- **W&B run:** n606gb5p (135 epochs, 360-min budget)
- **Command:** `CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --cosine-t-mult 2 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --agent mugen --wandb-name "mugen/tf-warm-restarts-t10-m2" --wandb-group "warm-restarts"`
- **Analysis:** Warm restarts found a 15.4% deeper basin at ep109 (25.46) vs fixed-period baseline at ep157 (30.10). The growing restart periods (10→20→40→... batches) provide progressively longer convergence windows within Lion's aggressive cosine cycling. All 4 val splits improved uniformly. Test metric at final epoch (36.29) is worse because the model isn't at a trough — test should be evaluated at best checkpoint.

### DrivAerML — All configurations WORSE ✗

| GPU | T_0 | T_mult | Best Val (%) | Epoch | Baseline | Change | W&B Run |
|-----|-----|--------|-------------|-------|----------|--------|---------|
| 7 | 10 | 3 | 5.969 | 83 | 4.619 | +29.2% | 93fo0uoh |
| 5 | 30 | 2 | 12.207 | 38 | 4.619 | +164.2% | zsglmzrs |
| 6 | 5 | 3 | 12.592 | 41 | 4.619 | +172.6% | gbiw8wa6 |
| 3 | 10 | 2 | 13.094 | 51 | 4.619 | +183.4% | 2kvavdu0 |
| 2 | 5 | 2 | 14.824 | 51 | 4.619 | +220.9% | 3n619bwv |
| 4 | 15 | 2 | 17.561 | 95 | 4.619 | +280.2% | ki6tf47p |

- **All commands used the DrivAerML golden config** with `--batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`
- Total epochs: 104-116 per run across the 360-min budget
- **Analysis:** DrivAerML's fixed-period CosineAnnealingLR at T_max=30 already provides excellent optimization — the baseline reached 4.619% at ep256. Warm restarts with growing periods mean LR stays at high values for longer stretches in later cycles, which appears destructive for DrivAerML's optimization landscape. The hard LR resets (jump to max) may disrupt the delicate gradient structure around the DrivAerML minima. T_mult=3 at T_0=10 (5.969%) was closest to baseline, likely because the longer periods allowed some convergence before restart. The DrivAerML runs only achieved 104-116 epochs (vs baseline's 267 in 180 min) — the warm restart disruptions may also slow convergence speed.

### What happened

**AirfRANS and TandemFoil:** Warm restarts with growing periods (T_mult=2) significantly improve over fixed-period cosine annealing. The mechanism: early short cycles (5-10 batches) enable aggressive exploration across the loss landscape, while later long cycles (40-80+ batches) allow the optimizer to settle into deeper basins. This matches the SGDR paper's findings — the restart schedule provides a natural curriculum from exploration to exploitation.

**DrivAerML:** The technique clearly hurts. DrivAerML's baseline already achieves deep convergence over 256+ epochs with fixed T_max=30. The hard LR restarts in warm restarts destroy accumulated optimization progress. DrivAerML's loss landscape may have narrower, more fragile basins that can't survive the periodic LR jumps to maximum.

**Key takeaway:** Warm restarts are dataset-dependent. They help when the baseline suffers from premature convergence (AirfRANS, TandemFoil), but hurt when the baseline already converges well (DrivAerML).

### Suggested follow-ups

1. **AirfRANS T_0=5, T_mult=2 with best-checkpoint evaluation** — The 0.000371 val result is remarkable but test was only evaluated at final epoch. Best-checkpoint saving would validate this is a real improvement.
2. **AirfRANS T_mult=3 at T_0=5** — Even longer later periods might help AirfRANS further.
3. **TandemFoil T_mult=3** — Similar to AirfRANS, longer later periods could push TandemFoil deeper.
4. **Do NOT pursue warm restarts for DrivAerML** — Clear negative result across 6 configurations.

**Comment by tcapelle** [comment]
ADVISOR: Thank you for this careful cross-dataset test. The direction is genuinely promising and I want to pursue it further.

**Verified results (W&B scan_history, best-epoch):**
| Dataset | Best Val (epoch) | Final Val | Baseline | Beat? |
|---|---|---|---|---|
| TandemFoil | 25.459 @ ep109 | 35.29 @ ep135 | 26.06 | YES (+2.3%) |
| AirfRANS | 0.000371 @ ep221 | 0.002385 @ ep351 | 0.000627 | YES (+40.8%) |
| DrivAerML | 5.969% @ ep83 | 8.969% @ ep106 | 4.619% | No |

**Critical issue: no best-checkpoint saving.** The warm-restart schedule creates sharp troughs in validation loss at each cycle nadir — the model genuinely reaches 0.000371 on AirfRANS and 25.459 on TandemFoil, but then the LR jumps back up and performance degrades. Since only the final checkpoint is saved, these results are unreachable. The AirfRANS transient win of 40.8% over baseline is a very strong signal worth capturing.

**Please re-run with these changes:**

1. **Add best-checkpoint saving.** The single most important change. Use early stopping or checkpoint-by-val-metric so the best-val model is preserved. Evaluate test metrics from that checkpoint. Without this, the best transient results are lost.

2. **TandemFoil / AirfRANS:** Keep T_0=10, T_mult=2 — this is clearly finding good solutions. Just add best-checkpoint so they can be captured.

3. **DrivAerML:** T_0=10/T_mult=2 and T_0=10/T_mult=3 both fail here. Try T_0=25, T_mult=1.5 or simply disable warm restarts (T_mult=1) for DrivAerML — it appears the longer DAM training dynamics don't suit short restart periods.

The 40.8% AirfRANS improvement at best-epoch is a real finding. Let's capture it properly.

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
